rust/src/container.rs

```diff
@@ -133,26 +141,77 @@ impl MappingBuilder {
     }
 }
 
-impl From<MappingBuilder> for ObjectMeta {
-    fn from(b: MappingBuilder) -> ObjectMeta {
+impl MappingBuilder {
+    /// Create an ObjectMeta for package-based content (excluding exclusive components)
+    fn create_package_meta(&self) -> ObjectMeta {
         let mut content = ObjectMetaMap::default();
-        for (checksum, paths) in b.checksum_paths {
-            // Use the first package name found for one of the paths (if multiple). These
-            // are held in sorted data structures, so this should be deterministic.
-            //
-            // If not found, use the unpackaged name.
-            let pkg = paths
+        for (checksum, paths) in &self.checksum_paths {
+            // Find content ID for this checksum, excluding components
+            let content_id = paths
+                .iter()
+                .filter_map(|p| self.path_packages.get(p))
+                .flatten()
+                .find(|id| {
+                    // Exclude component IDs - only use package IDs
+                    !self
+                        .componentmeta
+                        .iter()
```

Comment:
I think I commented on this before but there's an O(N²) thing going on here in that we're iterating over each file, and then for each file iterating over all componentmeta.

I feel like there has to be a way to clean this up a bit more.

Also can we deduplicate this via a helper function with what's in `create_component_meta`?

Reply:
ah, yea github collapsed your previous comments and I missed them this morning. I'll clean this up more.